### PR TITLE
refactor(test): dockerize tests and run sync in detached mode

### LINKED ISSUE #2411: Check that Google Cloud builds actually run the tests
## Motivation

There are a lot of command layers in the Google Cloud builds, so it's easy to accidentally skip tests or ignore failures.

Instead, we should:
- check that the test output contains successful cached sync test lines
- split long lines

And optionally:
- reduce the complexity or layers in the tests
- move the docker commands into a script file
  - refactor out common rebuild and test script steps

## Affected Files

https://github.com/ZcashFoundation/zebra/blob/19fa36049f345e76e81a94823b84f4e9b9a52a16/.github/workflows/test.yml#L67-L73

https://github.com/ZcashFoundation/zebra/blob/19fa36049f345e76e81a94823b84f4e9b9a52a16/.github/workflows/regenerate-stateful-test-disks.yml#L64-L70

## Related Work

This refactor should prevent bugs like #2402 and #2403.

### PR DESCRIPTION
**Pull request by gustavovalverde**
## Motivation
- Inherit changes from https://github.com/ZcashFoundation/zebra/pull/3431
- Improve machine types and disks for CPU-heavy workloads
- Make overall tests faster
- Workflows with `*patch.yml` that were not needed were causing a bunch of noise on GitHub dashboard and PRs
- Images were being build on GCP and then tested, this could be easily reuse images and cache from GitHub's Actions

Fixes: #2411 #3344 
Partially: #1592 #2653 #2995 #3447

## Solution
- Use C2 machines with SSD
- Change the state disk to SSD
- Deploy and run the sync on a single step (on-deploy) with gcloud, instead of deploying the machine, ssh into it, build, run and wait.
- Merge state regeneration and sync test on a single YAML
- Use the same Dockerfile for CD and test (another PR will merge both)
- Use a modular workflow with dependencies, reusing the same image for tests
- 

## Review
@teor2345 @conradoplg @dconnolly 

## Follow Up Work
- Merge `.build` and `.test` Dockerfiles
- Validate after a few tests if the boot disk size can be reduced
- https://github.com/ZcashFoundation/zebra/issues/2799
- Include changes from https://github.com/ZcashFoundation/zebra/pull/3543

### CODE REVIEW

File: .github/workflows/cd.yml
Comment (dconnolly) [inline .github/workflows/cd.yml]: `Mainnet` inside the Dockerfile unless overridden by manual deploy, right?

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml]:
```suggestion
        default: 'Mainnet'
```

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml]:
```suggestion
  NETWORK: Mainnet
```

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml]: ooooo 💻 💻 💻

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml:73]: 👍

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml:166]: How often do we want to do this?

File: .github/workflows/test.yml
Comment (dconnolly) [inline .github/workflows/test.yml]: Clever!

**Review by dconnolly — COMMENTED** [review]
Looking good, a few questions and small suggestions

File: .github/workflows/cd.yml
Comment (gustavovalverde) [reply .github/workflows/cd.yml]: Correct

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml]: Oh, yeah. I'll reduce that. There's no major impact between 4 and 16 CPU, the only impact is in the first 5-15 min of heavy computation, which creates a difference of around ~10 min overall. 

I'm not sure how much impact we're _expecting_ from having a bunch of CPUs but is not much, to be honest.

File: .github/workflows/test.yml
Comment (teor2345) [reply .github/workflows/test.yml:166]: Whenever we change the state database version?

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml:182]: These paths are ok for now, I've added a note to tickets #3150, #3151, and #3152 to update them if needed.

File: .github/workflows/test.yml
Comment (teor2345) [reply .github/workflows/test.yml:166]: (This is actually handled automatically by the change paths below this line.)

**Review by teor2345 — DISMISSED** [review]
I am ok with these changes once Deirdre's comments have been addressed.

**Review by teor2345 — CHANGES_REQUESTED** [review]
PR #3492 failed with:
> Error: google-github-actions/auth failed with: the GitHub Action workflow must specify exactly one of "workload_identity_provider" or "credentials_json"! If you are specifying input values via GitHub secrets, ensure the secret is being injected into the environment. By default, secrets are not passed to workflows triggered from forks, including Dependabot.

https://github.com/ZcashFoundation/zebra/runs/5118981346?check_suite_focus=true#step:4:58

It uses a branch from outside `ZcashFoundation/zebra`.

I'd like to fix that bug in PR #3431 before we merge this PR.

**Comment by teor2345** [comment]
I reviewed commit [refactor(test): seggregate disks regeneration from tests](https://github.com/ZcashFoundation/zebra/pull/3459/commits/cf2c3d17e2ac321c5de179f3388cf6a0945d9907), it seems to be ok, apart from this CI failure:
> Invalid value for field 'resource.disks[1].initializeParams.diskName': 'zebrad-cache-970b5d9-Mainnet-canopy'. Must be a match of regex '(?:[a-z](?:[-a-z0-9]***0,61***[a-z0-9])?)'

https://github.com/ZcashFoundation/zebra/runs/5119194528?check_suite_focus=true#step:6:90

But it's also a scope increase for this PR, which already got reviewed by 2 people.
Would you mind making any new changes in a separate PR?

**Comment by gustavovalverde** [comment]
@teor2345 when you refer to the new scope, I suppose you mean this commit: https://github.com/ZcashFoundation/zebra/pull/3459/commits/cf2c3d17e2ac321c5de179f3388cf6a0945d9907

The thing is that the previous state, before this commit, wasn't functional, as it was regenerating and then testing on a single run (12+ hours). I realized this after going through @dconnolly review.

**Comment by teor2345** [comment]
> @teor2345 when you refer to the new scope, I suppose you mean this commit: [cf2c3d1](https://github.com/ZcashFoundation/zebra/commit/cf2c3d17e2ac321c5de179f3388cf6a0945d9907)
> 
> The thing is that the previous state, before this commit, wasn't functional, as it was regenerating and then testing on a single run (12+ hours). I realized this after going through @dconnolly review.

Thanks, that makes sense - and thank you for testing and fixing this bug before we merge!

**Review by teor2345 — DISMISSED** [review]
Thanks for fixing those final bugs, I'm glad we're getting this merged.
I've been wanting to do something like this for months!

(I changed the priority to medium so that the failure fix in PR #3503 merges first, then this PR, then the docs PRs.)

**Comment by gustavovalverde** [comment]
I'm sorry that I'm making more changes, but it's actually fixing things we thought were ok.

And I'm also being hit by this issue when I'm trying to test this workflow while leveraging docker's cache. https://github.com/containerd/containerd/issues/5978

**Comment by teor2345** [comment]
> I'm sorry that I'm making more changes, but it's actually fixing things we thought were ok.
> 
> And I'm also being hit by this issue when I'm trying to test this workflow while leveraging docker's cache. [containerd/containerd#5978](https://github.com/containerd/containerd/issues/5978)

Thanks for fixing these, I'd rather catch them before we merge!

File: .github/workflows/test.yml
Comment (teor2345) [inline .github/workflows/test.yml:163]: Question: Why does this test need `-Zunstable-options`? Did we accidentally drop another unstable option?

https://doc.rust-lang.org/cargo/commands/cargo-build.html#common-options

I don't want to block the PR on this, happy to fix it up in a following PR.

**Comment by teor2345** [comment]
@gustavovalverde it looks like RAM is very low on the regenerate-state-disks VM. This might be happening because the disk performance is very slow. (Or the disk could be slow due to VM swapping.)

Do you want to try a `c2-standard-8` instead?

Here are the important parts of the failed logs:

<details>

Zebra starts downloading some new blocks, and checkpoints them:
```
Feb 14 23:10:06.609  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.750 remaining_sync_blocks=1469497
Feb 14 23:10:45.471  INFO ***net="Main"***:sync: zebrad::components::sync: extending tips tips.len=1 in_flight=2000 lookahead_limit=2000 state_tip=Some(Height(106374))
Feb 14 23:10:53.391  INFO ***net="Main"***:sync: zebrad::components::sync: waiting for pending blocks tips.len=1 in_flight=2473 lookahead_limit=2000
Feb 14 23:10:54.072  INFO ***net="Main"***:sync:extend_tips:checkpoint: zebra_consensus::checkpoint: verified checkpoint range block_count=340 current_range=(Excluded(Height(108533)), Included(Height(108873)))
```

It takes 4 minutes to write 700 blocks to disk:
(Normal speed is tens of thousands of blocks per minute on my machine.)
```
Feb 14 23:10:59.262  WARN ***net="Main"***:sync: zebrad::components::sync: error downloading and verifying block e=VerifierError(Elapsed(()))
Feb 14 23:10:59.262  INFO ***net="Main"***:sync: zebrad::components::sync: waiting to restart sync timeout=67s state_tip=Some(Height(106374))
Feb 14 23:11:06.610  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.750 remaining_sync_blocks=1469497
Feb 14 23:12:06.267  INFO ***net="Main"***:sync: zebrad::components::sync: starting sync, obtaining new tips state_tip=Some(Height(106374))
Feb 14 23:12:06.612  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.750 remaining_sync_blocks=1469498
Feb 14 23:13:06.613  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.750 remaining_sync_blocks=1469499
Feb 14 23:14:06.614  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.789 remaining_sync_blocks=1468878
Feb 14 23:15:06.616  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.798 remaining_sync_blocks=1468765
```

Zebra's Peer Set isn't getting requests, so it seems like it might be the disk that's slow:
```
Feb 14 23:15:06.850  INFO ***net="Main"***:crawl_and_dial: zebra_network::peer_set::inventory_registry: dropped lagged inventory advertisements count=41
```

Still writing blocks, 8 minutes to write 1200 blocks:
```
Feb 14 23:15:14.850  INFO ***net="Main"***:crawl_and_dial: zebra_network::peer_set::candidate_set: timeout waiting for peer service readiness or peer responses
Feb 14 23:16:06.617  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.802 remaining_sync_blocks=1468693
Feb 14 23:17:06.618  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.816 remaining_sync_blocks=1468453
Feb 14 23:18:06.619  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.821 remaining_sync_blocks=1468367
Feb 14 23:19:06.620  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.831 remaining_sync_blocks=1468221
Feb 14 23:20:06.621  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.839 remaining_sync_blocks=1468090
Feb 14 23:21:06.623  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.844 remaining_sync_blocks=1468007
Feb 14 23:22:06.624  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.855 remaining_sync_blocks=1467851
Feb 14 23:23:06.625  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.881 remaining_sync_blocks=1467450
```

The network part of the syncer got too far ahead, so it waited for a while:
```
Feb 14 23:23:29.400  INFO ***net="Main"***:sync:obtain_tips: zebrad::components::sync::downloads: synced block height too far ahead of the tip: dropped downloaded block. Hint: Try increasing the value of the lookahead_limit field in the sync section of the configuration file. hash=block::Hash("000000000039e5dfa40c8213067bcb288479559b595630f38f6eef36caaf261b") block_height=Height(1565894) tip_height=Some(Height(108873)) max_lookahead_height=Height(114873) lookahead_limit=2000
Feb 14 23:23:31.413  WARN ***net="Main"***:sync: zebrad::components::sync: error downloading and verifying block e=AboveLookaheadHeightLimit
Feb 14 23:23:31.413  INFO ***net="Main"***:sync: zebrad::components::sync: waiting to restart sync timeout=67s state_tip=Some(Height(108873))
Feb 14 23:24:06.627  INFO ***net="Main"***: zebrad::commands::start: estimated progress to chain tip sync_percent=6.909 remaining_sync_blocks=1467005
```

https://github.com/ZcashFoundation/zebra/runs/5192084284?check_suite_focus=true#step:9:1286

</details>

And then the process got killed for out of memory.

Happy to jump on a call or chat if that would help diagnose things. We're so close! 🤔

File: .github/workflows/test.yml
Comment (gustavovalverde) [reply .github/workflows/test.yml:163]: I'm adding this as this is being done with https://github.com/ZcashFoundation/zebra/blob/main/docker/Dockerfile.test#L30

Just to keep the same tests that have been running so far.

Comment (teor2345) [reply .github/workflows/test.yml:163]: Ah right sorry, I missed the `--include-ignored`, it might have been off my screen.

**Review by teor2345 — APPROVED** [review]
Looks great, let's merge!

**Pull request merged**
